Registry: Implement SOA query support

mkykadir opened the pull request:
Fixes #2448 

Adds support for SOA queries. Zone serial is updated when data is **really** updated or a record is removed from database.

Note:
Currently SOA query is not checking for Zone configuration if its authoritative, because current configuration parameters are not enough (secondaries are also authoritative, only resolvers are not authoritative)

Geod24 commented on source/agora/api/Registry.d:
```diff
@@ -30,6 +32,14 @@ public struct RegistryPayloadData
     /// monotonically increasing sequence number
     public ulong seq;
+
+    /// Compares payload data, ignores `seq` and permutations of `addresses` 
+    ///   are consiederd equal
+    bool opEquals (in RegistryPayloadData other) const @safe
+    {
+        return (this.public_key == other.public_key)
+            && (this.addresses.isPermutation(other.addresses));
+    }
```
If we implement `opEquals`, we generally implement `opCmp` as well.

Geod24 commented on source/agora/api/Registry.d:
```diff
@@ -30,6 +32,14 @@ public struct RegistryPayloadData
     public ulong seq;
+
+    /// Compares payload data, ignores `seq` and permutations of `addresses` 
+    ///   are consiederd equal
```
Why the extra indent ? Also typo.

Geod24 commented on source/agora/node/Registry.d:
```diff
@@ -928,6 +939,16 @@ private struct ZoneData
             return;
 
+        if (this.get(payload.payload.data.public_key) == payload)
+        { // Payload is equal to Zone's data, no need to update
```
Newline

Geod24 reviewed (approved):
This does three things that could be broken down in their own commits:
- Add an `opEqual` to a `struct`;
- Reply to `SOA` request;
- Avoid overriding payload data when not needed;

Concept LGTM.

Commit:
Overload `opEqual` and `opCmp` for registry payload

to consider payloads as equal when `addresses` are permutation of each
other

Commit:
Avoid zone update when payload is equal

since RFC#1034 requires SERIAL to be updated when zone is really changed.
This also prevents unnecessary SERIAL updates during periodic name registry.

Commit:
Add support to reply SOA questions

Geod24 commented on source/agora/api/Registry.d:
```diff
@@ -30,6 +33,48 @@ public struct RegistryPayloadData
+            
+            foreach (thisAddr, otherAddr; this.addresses.zip(other.addresses))
+            {
+                auto comparison = cmp(thisAddr.toString, otherAddr.toString);
```
Let's avoid `toString` as it allocates

Geod24 commented on source/agora/api/Registry.d:
```diff
@@ -30,6 +33,48 @@ public struct RegistryPayloadData
+            foreach (thisAddr, otherAddr; this.addresses.zip(other.addresses))
+            {
+                auto comparison = cmp(thisAddr.toString, otherAddr.toString);
+                // will eventually find non-equal addresses since opEquals failed
+                if (comparison == 0) continue;
+                return comparison;
```
```D
if (auto c = cmp(...))
    return c;
```

Geod24 commented on source/agora/node/Registry.d:
```diff
@@ -833,6 +842,7 @@ private struct ZoneData
             return Header.RCode.NameError;
 
         reply.answers ~= payload.toRR(question.qname);
+        reply.authorities ~= this.toRR(); // optional
```
I think most servers don't include this, but reading the RFC, this looks correct.

Geod24 reviewed (approved):
Just the allocation.

Geod24 reviewed (approved):
Merge at will
